### ISSUE TITLE
Scoreboard skip analysis: 656 onnx2burn codegen failures — ~200 are fixable config restrictions on supported ops

### ISSUE THREAD

**Issue by antimora** [tracel-ai/burn-onnx#314]
## Summary

When running the full ONNX backend test suite against burn-onnx (1678 node tests total), **656 tests (39%)** fail because `onnx2burn` panics during codegen. Crucially, these are **not all "not implemented" cases** — many are bugs, over-strict parser checks, and configuration restrictions on ops we already support (Clip, Slice, Split, ArgMax, ArgMin, Conv, ConvTranspose, AvgPool, MaxPool, TopK, OneHot, Dropout, Pad).

This corrects the inflated "90% pass rate" from #311. Once the scoreboard's `conftest.py` (via `harden_backend_test_skips`) converts `BackendIsNotSupposedToImplementIt` exceptions into real pytest skips, the honest breakdown is:

- **632 passed** (34.6%)
- **247 failed** (13.5%, mostly numerical/dtype errors already in #311)
- **948 skipped** (51.9%) — driven largely by the issues below

Full data: https://gist.github.com/antimora/c7373cd07204a9bd20295519c5404f20

## Failure categories

Grouped by root cause; counts are of distinct ONNX test cases hitting each.

### 1. Ops we already support but reject on specific configurations (~100+ tests)

These are the highest-value fixes. The op has a processor, but burn-onnx panics on a config variant it doesn't handle yet.

| Count | Issue | Example test |
|---:|---|---|
| 61 | `Unsupported Mod input types: lhs=Shape(1), rhs=Shape(1)` (crates/burn-onnx/src/burn/node/modulo.rs:112) | test_attention_3d_attn_mask_expanded |
| 19 | `LayerNorm: weight tensor must be present` (rejected when weight is a runtime input) | test_layer_normalization_2d_axis0 |
| 14 | `Flatten: input tensor must have at least 2 dimensions (got 1)` | test_layer_normalization_2d_axis1_expanded |
| 11 | `DequantizeLinear: Type mismatch: expected on-device tensor for x_scale, got ScalarNative(F32)` | test_dequantizelinear |
| 8 | `select_last_index=1 is not supported for argmax in burn` | test_argmax_default_axis_example_select_last_index |
| 8 | `select_last_index=1 is not supported for argmin in burn` | test_argmin_default_axis_example_select_last_index |
| 8 | `Clip: runtime min values are not supported in burn-onnx` | test_clip |
| 8 | `Runtime input is not implemented for Dropout` | test_dropout_default_mask_ratio |
| 8 | `Resize: custom axes attribute is not supported` | test_resize_downsample_sizes_nearest_not_smaller |
| 8 | `Split: Tensor for split sizes input / actual Shape(2)` | test_rotary_embedding_3d_input_expanded |
| 8 | `Slice: Axes must be provided by onnx-ir for tensor slice` | test_slice |
| 8 | `Split: Runtime split sizes are not supported in burn-onnx` | test_split_variable_parts_1d_opset13 |
| 6 | `Pad: axes input is not supported` | test_attention_4d_diff_heads_mask4d_padded_kv_expanded |
| 6 | `Only avg_pool 1d and 2d are supported` (3D avg pool) | test_averagepool_3d_default |
| 6 | `Conv2d: weight tensor must be present` (runtime weight) | test_basic_conv_with_padding |
| 6 | `ConvTranspose2d: weight tensor must be present` | test_convtranspose |
| 6 | `TopK: Runtime k value is not supported in burn-onnx` | test_top_k |
| 5 | `If: Squeeze cannot infer output rank when axes is None and input tensor static shape is unknown` | test_affine_grid_2d_align_corners_expanded |
| 4 | `Only max_pool 1d and 2d are supported` (3D max pool) | test_maxpool_3d_default |
| 4 | `OneHot with runtime depth is not supported in burn-onnx` | test_onehot_negative_indices |

**Subtotal: ~212 tests.** Each of these is a contained bug/restriction we could lift.

### 2. Type inference / IR bugs (~50 tests)

Type inference or IR phase fails for ops that should be handled.

| Count | Issue |
|---:|---|
| 19 | `Unknown node type: VariantNotFound` (op not registered in `Node` enum) |
| 3 | `DynamicQuantizeLinear: type mismatch` |
| 3 | `L1Normalization: type inference` |
| ... | Various other one-off type inference panics |

### 3. Over-strict parser check (~15 tests)

```
ONNX model must specify opset version for default domain
```
at `crates/onnx-ir/src/pipeline.rs:415`. Some test models (e.g. `test_adagrad`) only declare an opset for a non-default domain like `ai.onnx.preview.training`, and onnx-ir refuses to parse them. We could fall back to a default opset (or the latest known) when the default-domain opset is absent. 15 tests affected.

### 4. Genuinely unsupported ops (~270 tests)

These are ops `burn-onnx` has never supported. Most are either training, classical ML, DSP, or rare inference ops.

| Count | ONNX op |
|---:|---|
| 48 | `SoftmaxCrossEntropyLoss` (sce_*) |
| 28 | `NegativeLogLikelihoodLoss` (nllloss_*) |
| 19 | `RMSNormalization` |
| 10 | `Dft` |
| 9 | `ImageDecoder` |
| 8 | `RotaryEmbedding` |
| 8 | `NonMaxSuppression` |
| 8 | `QLinearMatMul` |
| 7 | `OptionalHasElement` |
| 7 | `TfIdfVectorizer` |
| ~130 | Other (ai.onnx.ml.*, string ops, sequence ops, training ops, etc.) |

### 5. Miscellaneous (~110 tests)

Exotic dtypes (cast/castlike FLOAT4E2M1, FLOAT8E*, INT2/4, UINT2/4), codegen panics in `burn_store` for certain weight initializers, etc.

## Why it matters

**Fixing category 1 alone (~212 tests) would move burn-onnx's scoreboard pass rate from ~35% to ~46%** without adding a single new operator. These are pure "we already handle this op but bail out on a specific config" cases.

Category 3 (15 tests) is a trivial one-line fix.

Categories 2 and parts of 1 are likely quick wins too (extending argmax/argmin to support `select_last_index=1`, Conv/ConvTranspose to accept runtime weights where possible, etc.).

Category 4 (~270 tests) is the real "roadmap" — these are ops we'd need to implement from scratch. But they're only **42% of the skips**, not the majority.

## Comparison with emx-onnx-cgen

emx-onnx-cgen sits at ~97% pass rate on the scoreboard with essentially zero skips. They raise plain `RuntimeError` instead of `BackendIsNotSupposedToImplementIt`, which counts as failures not skips, but the real difference is that they've actually implemented most of the ONNX operator set (their codegen covers ~98% of the test suite). Closing this gap is a multi-month roadmap item.

## Reproducing

```bash
cd scoreboard
ONNX2BURN=$(pwd)/../target/release/onnx2burn uv run python -c "
import os, subprocess, tempfile, json
import onnx

data_dir = os.path.join(os.path.dirname(onnx.__file__), 'backend/test/data/node')
fails = []
for name in sorted(os.listdir(data_dir)):
    p = os.path.join(data_dir, name, 'model.onnx')
    if not os.path.exists(p): continue
    with tempfile.TemporaryDirectory() as td:
        r = subprocess.run([os.environ['ONNX2BURN'], p, td, '--no-development', '--no-simplify'],
                           capture_output=True, timeout=30)
        if r.returncode != 0:
            fails.append(name)
print(f'{len(fails)} failures out of', len(os.listdir(data_dir)))
"
```

Refs #299, #311, #312

**Comment by antimora** [comment]
## Progress update — 39 tests unblocked in branch `issue/314-scoreboard-skip-fixes`

Working my way down the fixable-config-restriction list from the issue body. Six commits so far; every one verified end-to-end against the ONNX backend test suite (via `onnx2burn` -> cargo build -> run -> compare with the reference `.pb` tensors).

| Commit | Fix | Newly passing |
|---|---|---:|
| 80070e2f | Mod/And Shape inputs (attention expanded chain) | 0 direct (unblocks downstream) |
| efa0343e | ArgMax/ArgMin `select_last_index=1` | 16 |
| 1da5c998 | Clip runtime min/max | 7 |
| daaa198a | Dropout identity + mask output + default ratio | 6 |
| a2aa7f1d | OneHot runtime depth/values | 4 |
| d3b50fc7 | TopK runtime k | 6 |
| | **Total** | **39** |

All in the "config restriction on an op we already support" bucket from the issue — no new Burn primitives needed, just lifting overly strict checks and wiring up the runtime-input codegen path.

### Pattern that keeps showing up

Five of the six fixes are minor variations of the same "op X hard-rejects runtime input Y" bug. Each follows the same shape:

1. `scope.arg(runtime_arg)` to pull the runtime tensor into scope (preserving clone tracking).
2. A small prelude binding a `usize`/`f32` local via `.to_data().convert::<T>().as_slice::<T>().unwrap()[0]` (tensor form) or `arg_to_ident(arg)` + cast (ScalarNative form).
3. The prelude + main expression wrapped in a `let #output = { ... };` block so the temporaries (`__clip_min`, `__onehot_depth`, `__topk_k`, etc.) are scoped and can't collide across multiple instances of the same op in one forward().
4. The static path is preserved unchanged so existing snapshots don't get noisy block wrappers.

If we find ourselves writing this a fourth time, it's probably worth extracting a helper in `argument_helpers.rs` like:

```rust
fn runtime_scalar_to_native(
    arg: &Argument,
    target: DType,
    scope: &mut ScopeAtPosition<'_>,
) -> TokenStream
```

Worth keeping in mind as we tackle Split runtime sizes (8 more tests) and Pad axes input (6 more tests) next.

### The Mod/And commit didn't move the needle (yet)

80070e2f is real work but no tests flipped from fail to pass. Every attention-expanded test that hits Mod(Shape, Shape) also hits a chain of other bugs downstream: And(Shape, BoolTensor), then a Reshape panic with an empty-ident arg name. The fix is a prerequisite for a bigger attention-expanded unblock but doesn't stand alone. Worth filing a separate followup for the Reshape empty-ident issue once we get there.

### What's left in the contained-bugs bucket

From the issue body, still to do:

| Count | Issue |
|---:|---|
| 8 | Split runtime sizes (`crates/burn-onnx/src/burn/node/split.rs:27`) |
| 8 | Slice "Axes must be provided by onnx-ir" (`slice.rs:187`) |
| 6 | Pad axes input (`crates/onnx-ir/src/node/pad.rs:364`) |
| 6 | Conv/ConvTranspose weight tensor must be present |
| 14 | Flatten rank-1 input (`flatten input tensor must have at least 2 dimensions (got 1)`) |
| 11 | LayerNorm weight tensor must be present |
| 11 | DequantizeLinear `x_scale` ScalarNative handling |
| 6 | 3D avg/max pool |
| 5 | If subgraph squeeze rank inference |
| ~50 | Type-inference bugs (`VariantNotFound` etc.) |

And the big one: ~85 tests in the reduce op family (shape/keepdims/bool-inputs), which is a bigger refactor and probably deserves its own issue + PR.

### Next step

I'll open the PR with the 6 commits above as the first batch — contained, easy to review as a group, each commit is a single-op fix with a clear before/after. Will continue with Split/Pad/Flatten on a followup branch.

Note: #315 tracks the new `onnx-official-tests` crate which will become the regression gate for this kind of work; once it lands we'll be able to see these deltas in PR comments automatically instead of me hand-running the scoreboard harness.

**Comment by antimora** [comment]
## Status update — PR #316 opened, 10 commits total

The 6 commits from the previous update plus 4 review-response commits now live in **#316** (open for review):

| Commit | Fix | Unblocked (end-to-end) |
|---|---|---:|
| 80070e2f | Mod/And Shape inputs | 0 direct* |
| efa0343e | ArgMax/ArgMin `select_last_index=1` | 16 |
| 1da5c998 | Clip runtime min/max | 7 |
| daaa198a | Dropout identity + mask + default ratio | 6 |
| a2aa7f1d | OneHot runtime depth/values | 4 |
| d3b50fc7 | TopK runtime k | 6 |
| 02d8af50 | Gate Dropout noop elimination on single-output case (review) | bonus bug fix |
| d0fce21b | Comment accuracy + Shape mod fmod coverage (review) | +6 unit tests |
| a2185b49 | Apply cargo fmt (review) | - |
| 0b3f23c0 | Clip comment fix + OneHot i64 depth test (review) | +1 unit test |
| | **Total** | **39 end-to-end, +7 unit tests, +1 bug fix** |

\* Mod/And unblocks downstream panics in the attention-expanded chain but the same tests hit a Reshape-with-empty-ident bug next. Pre-req for a cascading fix.

## Reviews

Ran `/pr-review-toolkit:review-pr` twice against the PR. Agents caught:
- **1 critical blocker**: a stale unit test (`test_dropout_config_missing_input`) from my spec-default fix — resolved in 02d8af50.
- **1 latent pre-existing bug**: `DropoutProcessor::is_noop` unconditionally returned `true`, but noop-elimination only rewires `output[0]`, so the mask-output case would orphan the mask. Fixed in 02d8af50 by gating on `outputs.len() == 1`. This is a complementary fix since my codegen already handles the mask case.
- **5 important issues** (misleading comments, missing `fmod=true` coverage on Shape arms): fixed in d0fce21b.
- **3 Copilot comments** on the second round: the clip comment fix and new OneHot i64 depth test were valid; the OneHot f32→f64 suggestion was based on an incorrect assumption about Burn's `one_hot_fill` API (which takes concrete `f32`, not `ElementConversion`), so I reverted that part and documented the constraint.
- **8 latent pre-existing codegen correctness issues** that aren't regressions from this PR but deserve targeted fixes with test coverage — filed separately as **#317**.

## Running state of the contained-bugs roster

Crossed off in #316:

- [x] 16 ArgMax/ArgMin `select_last_index=1`
- [x] 7 Clip runtime min/max (of 10 hitting the panic; remaining 3 are int8 + `min_greater_than_max` edge case)
- [x] 6 Dropout runtime ratio + missing ratio + mask output
- [x] 4 OneHot runtime depth/values
- [x] 6 TopK runtime k (of 7; remaining is `test_top_k_smallest` on a separate parsing path)
- [x] Mod Shape inputs (codegen ok; cascades to next bug)
- [x] And Shape+BoolTensor (codegen ok; cascades)

Still to do (followup branches after #316 merges):

| Count | Issue |
|---:|---|
| 8 | Split runtime sizes (`crates/burn-onnx/src/burn/node/split.rs:27`) |
| 8 | Slice "axes must be provided" (`slice.rs:187`) |
| 6 | Pad axes input (`crates/onnx-ir/src/node/pad.rs:364`) |
| 6 | Conv/ConvTranspose runtime weights |
| 14 | Flatten rank-1 input |
| 11 | LayerNorm runtime weight |
| 11 | DequantizeLinear ScalarNative `x_scale` |
| 6 | 3D avg/max pool |
| 5 | If subgraph Squeeze rank inference |
| ~50 | Type-inference bugs (`VariantNotFound`, etc.) |
| ~85 | Reduce ops shape/keepdims (bigger refactor, separate PR) |

Plus the Reshape empty-ident cascade that blocks the 61 Mod/And attention-expanded tests.

## Related issues

- **#315** — `onnx-official-tests` native regression crate (so future batches like this show up as PR comment deltas automatically instead of me hand-running the scoreboard harness)
- **#317** — 8 latent codegen correctness issues surfaced by the #316 review

**Comment by antimora** [comment]
Superseded by `crates/onnx-official-tests`, which now tracks every vendored node test with a status and reason in `expectations.toml` and fails CI on drift. As of fedcb0b7 the numbers are 1139 pass, 501 skip-codegen, 77 skip-compile, 42 fail-compare, 6 flaky. The per-bucket follow-ups filed from here are either fixed (#345, #346, #352) or still tracked on their own (#344, #348, #350, #351).
